File: SWEA/D6/1267. ［S／W 문제해결 응용］ 10일차 － 작업순서/［S／W 문제해결 응용］ 10일차 － 작업순서.py
from collections import deque

def topological_sort():
  result = []
  queue = deque([])

  # 만약 진입차수가 0이면 queue에 노드 추가
  for i in range(1, v + 1):
    if indegree[i] == 0:
      queue.append(i)

  while queue:
    # 큐에서 노드 제거
    x = queue.popleft()

    # 출력 배열에 저장
    result.append(x)

    # 연결된 노드의 간선 제거
    for i in graph[x]:
      indegree[i] -= 1

      # 진입 차수가 0이 되면 큐에 해당 노드 추가
      if indegree[i] == 0:
        queue.append(i)
  
  return result


for tc in range(1, 11):
  v, e = map(int, input().split())
  lst = list(map(int, input().split()))

  graph = [[] for _ in range (v + 1)]
  indegree = [0] * (v + 1)

  # 인접그래프 만들면서 노드별 진입차수 기록
  for i in range(e):
    start = lst[i * 2]
    end = lst[i * 2 + 1]
    graph[start].append(end)
    indegree[end] += 1

  # 위상 정렬 실행
  res = topological_sort()

  print(f"#{tc}", end = ' ')
  print(*res)

File: SWEA/D6/1267. ［S／W 문제해결 응용］ 10일차 － 작업순서/［S／W 문제해결 응용］ 10일차 － 작업순서.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
#include <stack>
#include <cstring>
using namespace std;

#define endl '\n'

typedef long long ll;


int v, e;
vector<int> graph[1001];
int indegree[1001];
vector<int> res;

void topologicalSort() {
    queue<int> q;

    for (int i = 1; i < v + 1; ++i) {
        if (indegree[i] == 0) {
            q.push(i);
        }
    }

    while (!q.empty()) {
        int x = q.front();
        q.pop();

        res.push_back(x);

        for (auto i: graph[x]) {
            indegree[i]--;

            if (indegree[i] == 0) {
                q.push(i);
            }
        }
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    for (int tc = 1; tc <= 10; ++tc) {
        memset(indegree, 0, sizeof(indegree));
        res.clear();
        cin >> v >> e;

        for (int i = 0; i <= v; ++i) {
            graph[i].clear();
        }

        for (int i = 0; i < e; ++i) {
            int start, end;
            cin >> start >> end;
            graph[start].push_back(end);
            indegree[end] += 1;
        }

        topologicalSort();
        cout << "#" << tc << " ";
        for (auto &iter: res) {
            cout << iter << " ";
        }
        cout << endl;
    }
}
